javalib/src/main/scala/java/lang/Throwables.scala

```diff
@@ -49,49 +49,126 @@ private[lang] object StackTrace {
   }
 }
 
-class Throwable(s: String, private var e: Throwable)
+class Throwable protected (s: String,
+                           private[this] var e: Throwable,
+                           enableSuppression: scala.Boolean,
+                           writableStackTrace: scala.Boolean)
     extends Object
     with java.io.Serializable {
+
+  def this(message: String, cause: Throwable) =
+    this(message, cause, true, true)
+
   def this() = this(null, null)
+
   def this(s: String) = this(s, null)
+
   def this(e: Throwable) = this(if (e == null) null else e.toString, e)
 
-  private var stackTrace: Array[StackTraceElement] = _
+  private[this] var stackTrace: Array[StackTraceElement] = _
 
-  fillInStackTrace()
+  if (writableStackTrace)
+    fillInStackTrace()
 
-  def initCause(cause: Throwable): Throwable = {
-    e = cause
-    this
-  }
+  // We use an Array rather than, say, a List, so that Throwable does not
+  // depend on the Scala collections.
+  private[this] var suppressed: Array[Throwable] = _
 
-  def getMessage(): String = s
+  final def addSuppressed(exception: Throwable): Unit = {
 
-  def getCause(): Throwable = e
+    if (exception == null) {
+      // Yes, JVM message uses 'cannot' (sic) & has terminal period/full_stop!
+      throw new java.lang.NullPointerException(
+        "Cannot suppress a null exception.")
+    }
 
-  def getLocalizedMessage(): String = getMessage()
+    if (exception == this) {
```

Comment:
This test should use `eq`. Otherwise you're going to call `equals` for every `addSuppressed`, and that does not match the behavior of the JVM.

Reply:
line #85 changed as recommended.

Based on this review comment, I swept through the code. In the places where the
js original (from a couple of days ago) used eq, I now use eq.   In places where I added code, I now use eq.

The java original had 12 instances of `== null` and 7 instances of `eq null`.
The former seemed like they all could/should have been `eq`.  From lack of
sufficient understanding, left those 12 as `==`  in this edit.

Please advise.  Thank you.

Reply:
If one of the sides is a *literal* `null`, `==` and `eq` are equivalent. So this is fine.